### ISSUE TITLE
Describe when calculates, constraints and relevancy are recomputed

### ISSUE THREAD

**Issue by lognaturel** [opendatakit/docs#213]
This will require a fairly deep understanding of Collect and JavaRosa and/or experimentation.

**Comment by lognaturel** [comment]
I think thinks will go in two places. One narrative for users in a section on form design and one narrative for developers about the JR innards.

For users I'm thinking of info like:
- When a value changes, all dependent nodes are recomputed
- When a form is finalized, all values are recalculated from start to finish
- The `once` function can be used to only evaluate an expression if the node's value is empty (e.g. `once(random())`)

For devs:
- `Triggerable` represents an expression, its context and the nodes that should be updated when the value of the expression changes...
- A directed acyclic graph of triggerables is kept and manages the recomputation processes...

@mdudzinski since you've been spending time in this space recently, it would be great to get an notes you have as comments on this issue.

**Comment by mdudzinski** [comment]
@lognaturel Sure! I will focus on the dev stuff first.

**Comment by mdudzinski** [comment]
@opendatakit-bot claim

**Comment by lognaturel** [comment]
#413 is awaiting a decision on where it will actually go.

**Comment by adammichaelwood** [comment]
This is covered now in Form Logic:

https://docs.opendatakit.org/form-logic/#when-expressions-are-evaluated

**Comment by sebmercier** [comment]
Hello,

This is my first contribution here, from a freshly created GitHub account, so I hope this the right place (and procedure) to post my comment. Please let me know if that's not the case or if I should create a new issue for this.

I'm precisely coming from the link posted by @adammichaelwood ( https://docs.opendatakit.org/form-logic/#when-expressions-are-evaluated ), because I realised that the `now()` function doesn't behave (anymore?) as described there. Used on its own (as in the "datetime_last_saved" example), it is evaluated only once when the survey is opened. I've read several discussions here about this subject (and related ones) over the past few weeks and I'm not sure where the debate is at today. But because the ODK documentation is an important source of information for XLSForm coders such as me, I thought I should point out this problem. (BTW, at the end of this same documentation section, the calculation formula should read `if(${age} = '', '', once(now()))` instead of `if(age = '', '', once(now()))`. I wasted more time than I care to admit because of this, back in December, but really, that's on me. :-) )

Here is some context, in case you'd have some guidance on the particular problem I'm facing. We're suspecting massive "Interviewer Fabrication" (to quote the title of Benjamin Birnbaum's dissertation, linked in another discussion by @lognaturel) in some of our recent assessments, initially based on how quickly the surveys were filled. So to be able to monitor this in the future, we'd like to calculate the duration of each interview _in-survey_. The reasons for using this instead of the more elaborate "audit" feature (which I haven't tested yet) are 1) to keep it light, and 2) to give the survey managers an easy way to monitor this indicator on the KoBo online interface, even mid-assessment, so they can take measures if necessary.

My first implementation of this was to use a "calculate" variable with `${end}-${start}` (and some time unit conversion). It worked really well in my first tests with Enketo, but I realised a few days ago that ODK/KoBo Collect behave differently: variables using `${end}` are not re-evaluated when the survey is finalised and their values remain empty.

Yesterday, I finally came up with another solution: I'm still using `${start}` as the initial timestamp, but I'm now using `if(${some_variable}='','',now())` for the final timestamp, again with some time unit conversion. I'm not sure why exactly (the behaviour of `now()` is still a bit of a mystery to me), but this seems to work as I want it to. (See attached XLS form: [Convert timestamps and calculate durations within survey.xlsx](https://github.com/opendatakit/docs/files/2763986/Convert.timestamps.and.calculate.durations.within.survey.xlsx)) However, I still have two problems with this solution:

1. I don't really know why it works and how `now()` works, so I'm afraid it may not work anymore in the future if the behaviour of `now()` changes at some point.
2. It uses a reference to an unrelated variable (which, in addition, needs to be **required** and **always relevant**), which may need to be changed when coding a different survey: I'd like to find a more self-contained solution.

Any insight into this matter would be greatly appreciated.

**Comment by sebmercier** [comment]
Done! Thanks for the suggestion, @yanokwa!
